### PR TITLE
Make focus command and goal selection for query commands independent of ltac1

### LINKED ISSUE #12004: Shared proof mode grammar
As of today, bullets and focusing brackets (and commands) and some proof-mode only commands (like `Grab Existential Variables`) are defined in the main grammar and thus can be run outside proof-mode (which makes no sense). OTOH, goal selectors, including goal selectors in front of focusing brackets, are only defined in the Ltac1 grammar, and thus cannot be used from Ltac2 (until they are re-implemented in Ltac2).

It seems to me that all of these should be defined in a special grammar that is loaded by default by all proof modes (unless they specifically reject it). So that, for instance, `{` is rejected outside of proof editing mode, and that `2: {` is accepted by default by Ltac2.

EDIT: the discussion below from Oct. 2020 is on a slightly different topic which is whether parsing proof commands / tactics in all context (and failing at interpretation time) could lead to error messages that are easier to understand.

### LINKED ISSUE #18351: Ltac2 does not support focus command with goal selector (`1:{` etc)
~~~coq
Require Import Ltac2.Ltac2.

Goal True.
  1:{ exact I. }
Qed.
~~~

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #18351
Fix #12004 

Overlays:
- https://github.com/Mtac2/Mtac2/pull/399

### CODE REVIEW

**Comment by SkySkimmer** [comment]
It doesn't factorize properly, not sure how to fix.
There's `query_command`, `{` and arbitrary tactics which can be preceded with a toplevel selector. Arbitrary tactics also support `par:`, and their grammar is only active in proof mode.

File: plugins/ltac/g_ltac.mlg [hunk 1]
```diff
@@ -282,10 +284,10 @@ GRAMMAR EXTEND Gram
   tactic:
     [ [ tac = ltac_expr -> { tac } ] ]
   ;
-
   tactic_mode:
-    [ [ g = OPT toplevel_selector; tac = G_vernac.query_command -> { Vernacexpr.VernacSynPure (tac g) }
-      | g = OPT toplevel_selector; "{" -> { Vernacexpr.VernacSynPure (Vernacexpr.VernacSubproof g) } ] ]
+    [ [ p = subprf -> { Vernacexpr.VernacSynPure p }
+      | g = OPT toplevel_selector; p = subprf_with_selector -> { Vernacexpr.VernacSynPure (p g) }
+      ] ]
```
Comment (SkySkimmer) [inline plugins/ltac/g_ltac.mlg:290]: Maybe adding these 2 rules could/should be automated in register_proof_mode? Not in this PR though I think.

**Comment by SkySkimmer** [comment]
I think it works now, although the proof modes need to manually coordinate so eg it won't automatically work with mtac2 (but should be easy to adapt mtac2?)

**Comment by SkySkimmer** [comment]
Test suite failure is because an interp time error became a parsing error (which can't be caught by Fail). We just need to turn the test into an output test.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**